packages/commonwealth/server/controllers/server_groups_methods/refresh_community_memberships.ts

```diff
@@ -136,3 +149,107 @@ export async function __refreshCommunityMemberships(
     } within ${(Date.now() - startedAt) / 1000}s`,
   );
 }
+
+export function makeGetBalancesOptions(
```

Comment:
Can we not retrieve all this data using SQL queries directly rather than looping through everything again?

Reply:
This is a pure compute function that takes the community's groups/addresses and outputs a list of options used for batch operations. No SQL involved, done within milliseconds.

The DB fetching happens in the caller scope, and it's already fairly efficient since it fetches all groups and addresses upfront without loops.

Reply:
What I meant was is there a reason we can't modify the caller scope fetch operation to return the data all neatly formatted? Not introducing loops or anything like that, on the contrary, removing loops and this function entirely by fetching the data in the format that we need right out of the gate. Seems like a doable SQL query to select just the group requirements mapped to appropriate addresses.

Reply:
I'm all for DDD as long as we aren't sacrificing performance or increasing logical complexity for it. If a single raw query can do both (reduce complexity AND improve performance) that should be something we consider. Suppose 3 requirements per group with 15 groups and 10k addresses in a single community. That's 450,000 iterations of the loop in the above function no? All that could be simplified to a single query:

```
SELECT
    DISTINCT ON (source_data.source_type, source_data.source)
    source_data.source_type AS balanceSourceType,
    source_data.source,
    array_agg(DISTINCT a.address) AS addresses
FROM
    (SELECT
         g.id AS group_id,
         jsonb_extract_path_text(jsonb_array_elements(g.requirements)->'data'->'source', 'source_type') AS source_type,
         jsonb_array_elements(g.requirements)->'data'->'source' AS source
     FROM
         "Groups" g
    ) AS source_data
        JOIN
    "Memberships" m ON source_data.group_id = m.group_id
        JOIN
    "Addresses" a ON m.address_id = a.id
GROUP BY
    source_data.source_type, source_data.source
ORDER BY
    source_data.source_type, source_data.source;
```

This is probably not hugely important for shipping gating as is but it's worth keeping in mind as gating scales and while as you say we move toward DDD + better test coverage.

Reply:
That's great. It's succinct and could work.

Due to time constraints, let's revisit this once we've got everything in working condition and can start measuring.

Reply:
FYI: I tested the query locally and the result shape is correct, but it’s not returning all sources. Probably related to the fact that we need to also include addresses that don't have associated memberships yet.

Wanted to test the perf, but I don't want to get too stuck in the weeds right now. Just a note for when we revisit.

Reply:
Made a ticket for this: https://github.com/hicommonwealth/commonwealth/issues/5809